NOISSUE - Introduce NATS Jetstream as Default ES

rodneyosodo opened the pull request:
### What does this do?
This PR introduces Nats Jetstream as the default event sourcing (ES) mechanism.

### Which issue(s) does this PR fix/relate to?
No issue

### List any changes that modify/break current functionality
This PR modifies mainflux default event sourcing mechanism to use Nats Jetstream instead of redis. This could potentially break existing code that relies on the previous ES implementation.

### Have you included tests for your changes?
Yes, this PR includes comprehensive tests to ensure that the Nats Jetstream implementation functions correctly and does not introduce regressions.

### Did you document any new/modified functionality?

### Notes
To be merged after #1897 , #1904 , #1893 , #1896 
Jetstream documentation - https://docs.nats.io/nats-concepts/jetstream/

rodneyosodo committed:
Use redis URL to configure username, password and db

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>
Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>
Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Refactor message broker implementation

This commit refactors the Nats message broker implementation to include pubsub options. These changes include:

- Adding `Option` func that takes in the URL and prefix
- Implement `WithStream` option which can create a different stream for nats stream
- Implement `WithExchange` option which can create a different exchaange for rabbitmq channel
- Implement `WithPrefix` option which allows to you change the publisher prefix

These changes improve the organization and readability of the codebase.

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>
Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>
Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Make event store configurable on dev deployment

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>
Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>
Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Fix adds options to messaging `PubSub` interface

Adding options to PubSub interface allows the use of messaging
package to do es.

The changes in this commit ensure that the code handles errors
properly and provides more informative error messages when
encountering unexpected types.

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>
Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>
Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Add NATS event publisher implementation

This commit adds the implementation of the NATS event publisher.

The NATS event publisher is responsible for publishing events
to a NATS messaging system. It uses the `messaging`
package to interact with the messaging system.

The implementation includes the following features:

- Publishing events to NATS using the `Publish` method.
- Marshaling events to JSON before publishing.
- Setting the message subject and headers based on the event.
- Handling errors during publishing.

This implementation is built with the `!rabbitmq` build tag,
which means it will only be compiled if the `rabbitmq` build tag
is not present.

The NATS event publisher is part of the Mainflux events package
and provides support for the Mainflux NATS events source service functionality.

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>
Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>
Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Add RabbitMQ event publisher implementation

This commit adds the implementation of the RabbitMQ event publisher.

The RabbitMQ event publisher is responsible for publishing events
to a RabbitMQ messaging system. It uses the `messaging`
package to interact with the messaging system.

The implementation includes the following features:

- Publishing events to RabbitMQ using the `Publish` method.
- Marshaling events to JSON before publishing.
- Setting the message subject and headers based on the event.
- Handling errors during publishing.

This implementation is built with the `rabbitmq` build tag,
which means it will only be compiled if the `rabbitmq` build tag
is present.

The RabbitMQ event publisher is part of the Mainflux events package
and provides support for the Mainflux RabbitMQ events source service functionality.

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>
Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>
Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Add configurable implementation for events store

This commit adds a new file `brokers_*.go` which contains the implementation for the different event store. The file includes functions for creating a new publisher and subscriber using different es store.

This commit also includes an `init` function that logs a message indicating that the binary was built using the respective package as the events store.

The purpose of this commit is to add support for alternative events store.

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>
Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>
Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Fix build flags

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>
Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>
Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Refactor Makefile and Semaphore configuration

The Makefile has been refactored to include the `MF_ES_STORE_TYPE` tag in the `go build` command. Additionally, the Semaphore configuration has been updated to include a new task for compiling with Redis as the broker type.

This commit addresses the need to compile the codebase with Redis as the event store type and includes the necessary changes in the Makefile and Semaphore configuration.

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>
Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>
Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Reduced due to memory on testing

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>
Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>
Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Fix tests for es

Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>
Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>
Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

SammyOina commented on pkg/events/nats/publisher.go:
```diff
@@ -0,0 +1,90 @@
+
+func (es *pubEventStore) StartPublishingRoutine(ctx context.Context) {
+	// Nats doesn't need to check for unpublished events
+	// since the events a publisher to a buffer.
```
```suggestion
	// since the events are published to a buffer.
```

rodneyosodo committed:
Fix grammar

Co-authored-by: Sammy Kerata Oina <44265300+SammyOina@users.noreply.github.com>
Signed-off-by: Rodney Osodo <socials@rodneyosodo.com>
Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>
Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Fix linting

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>
Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
feat(docker): update environment variables for message broker

The commit updates the environment variable `MF_ES_STORE_TYPE` in the `docker/.env` file. The variable is changed from `${MF_MQTT_BROKER_TYPE}` to `${MF_MESSAGE_BROKER_TYPE}` to accurately reflect the type of message broker being used. This change ensures that the correct message broker is configured for the Event Store.

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>
Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
feat: Update docker environment variables

- Removed the unused MF_ES_URL variable in the .env file
- Updated the MF_ES_STORE_TYPE and MF_ES_STORE_URL variables in the .env file to match the MF_MESSAGE_BROKER_TYPE and MF_NATS_URL variables respectively

Signed-off-by: Rodney Osodo <28790446+rodneyosodo@users.noreply.github.com>
Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
Fix after rebase

Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

dborovcanin commented on pkg/messaging/pubsub.go:
```diff
@@ -59,3 +59,6 @@ type PubSub interface {
 }
+
+// Option represents optional configuration for message broker.
+type Option func(vals interface{}) error
```
Please add a little more details to the description explaining why we have this type and how it's meant to be used. Overall, the idea is good and generic and I like it, but it can get a little tricky to understand it from the code and a comment with a better explanation will help. Consider adding an example comment, as well.

rodneyosodo committed:
Add godocs for option parameters for brokers

Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

SammyOina commented on pkg/messaging/nats/options.go:
```diff
@@ -0,0 +1,56 @@
+var ErrInvalidType = errors.New("invalid type")
+
+// WithPrefix sets the prefix for the publisher.
+func WithPrefix(prefix *string) messaging.Option {
```
With prefix for function names is typically used when extending some receiver value (usually a struct) consider renaming this

rodneyosodo committed:
pass by value exchange and prefix names

Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

dborovcanin commented on pkg/messaging/rabbitmq/pubsub.go:
```diff
@@ -15,14 +15,13 @@ import (
+const SubjectAllChannels = "channels.#"
 
 var (
+	exchangeName = "messages"
+	chansPrefix  = "channels"
```
Use `const`.

rodneyosodo committed:
Rename option functions

Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

SammyOina commented on pkg/messaging/pubsub.go:
```diff
@@ -59,3 +59,22 @@ type PubSub interface {
+//
+// Example:
+//
+//	broker.NewPublisher(ctx, url, broker.WithPrefix(&eventsPrefix), broker.WithJSStream(js))
```
update comment

rodneyosodo committed:
move variables to constants

Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

rodneyosodo committed:
fix: option example comment

Signed-off-by: rodneyosodo <blackd0t@protonmail.com>

drasko reviewed (approved):
LGTM
